<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Kasper D. Hansen" />


<title>rtracklayer - Data Import</title>


<style type="text/css">code{white-space: pre;}</style>
<link href="data:text/css;charset=utf-8,pre%20%2Eoperator%2C%0Apre%20%2Eparen%20%7B%0Acolor%3A%20rgb%28104%2C%20118%2C%20135%29%0A%7D%0Apre%20%2Eliteral%20%7B%0Acolor%3A%20%23990073%0A%7D%0Apre%20%2Enumber%20%7B%0Acolor%3A%20%23099%3B%0A%7D%0Apre%20%2Ecomment%20%7B%0Acolor%3A%20%23998%3B%0Afont%2Dstyle%3A%20italic%0A%7D%0Apre%20%2Ekeyword%20%7B%0Acolor%3A%20%23900%3B%0Afont%2Dweight%3A%20bold%0A%7D%0Apre%20%2Eidentifier%20%7B%0Acolor%3A%20rgb%280%2C%200%2C%200%29%3B%0A%7D%0Apre%20%2Estring%20%7B%0Acolor%3A%20%23d14%3B%0A%7D%0A" rel="stylesheet" type="text/css" />
<script src="data:application/x-javascript;base64,
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();

"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>


<link href="data:text/css;charset=utf-8,body%2C%20td%20%7B%0Afont%2Dfamily%3A%20sans%2Dserif%3B%0Abackground%2Dcolor%3A%20white%3B%0Afont%2Dsize%3A%2013px%3B%0A%7D%0Abody%20%7B%0Amax%2Dwidth%3A%20800px%3B%0Amargin%3A%200%20auto%3B%0Apadding%3A%201em%201em%202em%3B%0Aline%2Dheight%3A%2020px%3B%0A%7D%0A%0Ap%2C%20pre%20%7B%20margin%3A%200em%200em%201em%3B%0A%7D%0A%0Aimg%2C%20table%20%7B%0Amargin%3A%200em%20auto%201em%3B%0A%7D%0Ap%20%7B%0Atext%2Dalign%3A%20justify%3B%0A%7D%0Att%2C%20code%2C%20pre%20%7B%0Afont%2Dfamily%3A%20%27DejaVu%20Sans%20Mono%27%2C%20%27Droid%20Sans%20Mono%27%2C%20%27Lucida%20Console%27%2C%20Consolas%2C%20Monaco%2C%20monospace%3B%0A%7D%0Ah1%2C%20h2%2C%20h3%2C%20h4%2C%20h5%2C%20h6%20%7B%20font%2Dfamily%3A%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Amargin%3A%201%2E2em%200em%200%2E6em%200em%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%20250%25%3B%0Afont%2Dweight%3A%20normal%3B%0Acolor%3A%20%2387b13f%3B%0Aline%2Dheight%3A%201%2E1em%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%20160%25%3B%0Afont%2Dweight%3A%20normal%3B%0Aline%2Dheight%3A%201%2E4em%3B%0Aborder%2Dbottom%3A%201px%20%231a81c2%20solid%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%20130%25%3B%20%7D%0Ah2%2C%20h3%20%7B%0Acolor%3A%20%231a81c2%3B%0A%7D%0Ah4%2C%20h5%2C%20h6%20%7B%0Afont%2Dsize%3A115%25%3B%0A%7D%20%0A%0Aa%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Aactive%20%7B%20outline%3A%20none%3B%20%7D%0Aa%3Avisited%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Ahover%20%7B%20color%3A%20%234c94c2%3B%20%7D%0Apre%2C%20img%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Adisplay%3A%20block%3B%0A%7D%0Apre%20%7B%0Aborder%3A%200px%20none%3B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0Awhite%2Dspace%3A%20pre%3B%0Aoverflow%2Dx%3A%20auto%3B%0A%7D%0Apre%20code%20%7B%0Aborder%3A%201px%20%23aaa%20dashed%3B%0Abackground%2Dcolor%3A%20white%3B%0Adisplay%3A%20block%3B%0Apadding%3A%201em%3B%20color%3A%20%23111%3B%0Aoverflow%2Dx%3A%20inherit%3B%0A%7D%0A%0Apre%20code%5Bclass%5D%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Apre%5Bclass%5D%20code%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Acode%20%7B%20background%2Dcolor%3A%20transparent%3B%0Acolor%3A%20%2387b13f%3B%0Afont%2Dsize%3A%2092%25%3B%0A%7D%0A%0Atable%2C%20td%2C%20th%20%7B%0Aborder%3A%20none%3B%0Apadding%3A%200%200%2E5em%3B%0A%7D%0A%0Atbody%20tr%3Anth%2Dchild%28odd%29%20td%20%7B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0A%7D%0Ablockquote%20%7B%0Acolor%3A%23666666%3B%0Amargin%3A0%3B%0Apadding%2Dleft%3A%201em%3B%0Aborder%2Dleft%3A%200%2E5em%20%23EEE%20solid%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200px%3B%0Aborder%2Dbottom%3A%20none%3B%0Aborder%2Dtop%2Dwidth%3A%20thin%3B%0Aborder%2Dtop%2Dstyle%3A%20dotted%3B%0Aborder%2Dtop%2Dcolor%3A%20%23999999%3B%0A%7D%0A%40media%20print%20%7B%0A%2A%20%7B%0Abackground%3A%20transparent%20%21important%3B%0Acolor%3A%20black%20%21important%3B%0Afilter%3Anone%20%21important%3B%0A%2Dms%2Dfilter%3A%20none%20%21important%3B%0A%7D%0Abody%20%7B%0Afont%2Dsize%3A12pt%3B%0Amax%2Dwidth%3A100%25%3B%0A%7D%0Aa%2C%20a%3Avisited%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Avisibility%3A%20hidden%3B%0Apage%2Dbreak%2Dbefore%3A%20always%3B%0A%7D%0Apre%2C%20blockquote%20%7B%0Apadding%2Dright%3A%201em%3B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Atr%2C%20img%20%7B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%20%21important%3B%0A%7D%0A%40page%20%3Aleft%20%7B%0Amargin%3A%2015mm%2020mm%2015mm%2010mm%3B%0A%7D%0A%40page%20%3Aright%20%7B%0Amargin%3A%2015mm%2010mm%2015mm%2020mm%3B%0A%7D%0Ap%2C%20h2%2C%20h3%20%7B%0Aorphans%3A%203%3B%20widows%3A%203%3B%0A%7D%0Ah2%2C%20h3%20%7B%0Apage%2Dbreak%2Dafter%3A%20avoid%3B%0A%7D%0A%7D%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">rtracklayer - Data Import</h1>
<h4 class="author"><em>Kasper D. Hansen</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#dependencies">Dependencies</a></li>
<li><a href="#overview">Overview</a></li>
<li><a href="#other-resources">Other Resources</a></li>
<li><a href="#the-import-function">The import function</a></li>
<li><a href="#bed-files">BED files</a></li>
<li><a href="#bigwig-files">BigWig files</a></li>
<li><a href="#other-file-formats">Other file formats</a></li>
<li><a href="#extensive-example">Extensive example</a></li>
<li><a href="#liftover">LiftOver</a></li>
<li><a href="#importing-directly-from-ucsc">Importing directly from UCSC</a></li>
<li><a href="#tabix-indexing">Tabix indexing</a></li>
<li><a href="#sessioninfo">SessionInfo</a></li>
</ul>
</div>

<div id="dependencies" class="section level2">
<h2>Dependencies</h2>
<p>This document has the following dependencies:</p>
<pre class="r"><code>library(rtracklayer)
library(AnnotationHub)
library(Rsamtools)</code></pre>
<p>Use the following commands to install these packages in R.</p>
<pre class="r"><code>source(&quot;http://www.bioconductor.org/biocLite.R&quot;)
biocLite(c(&quot;rtracklayer&quot;, &quot;AnnotationHub&quot;, &quot;Rsamtools&quot;))</code></pre>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>The <em><a href="http://bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a></em> package interfaces to (UCSC) Genome Browser. It contains functions for importing and exporting data to this browser.</p>
<p>This includes functionality for parsing file formats associated the UCSC Genome Browser such as BED, Wig, BigBed and BigWig.</p>
</div>
<div id="other-resources" class="section level2">
<h2>Other Resources</h2>
<ul>
<li>The vignette from the <a href="http://bioconductor.org/packages/rtracklayer">rtracklayer webpage</a>.</li>
</ul>
</div>
<div id="the-import-function" class="section level2">
<h2>The import function</h2>
<p>The function to parse data formats is <code>import()</code>. This function has a <code>format</code> argument taking values such as <code>BED</code> or <code>BigWig</code>.</p>
<p>Note that there is a help page for the general <code>import()</code> function, but there are also file format specific help pages. The easiest way to get to these help pages is to look for <code>XXFile</code> with <code>XX</code> being the format.</p>
<pre class="r"><code>?import
?BigWigFile</code></pre>
<p>There are often format specific arguments.</p>
</div>
<div id="bed-files" class="section level2">
<h2>BED files</h2>
<p>Most BED files are small and can be read as a single object. The output of <code>import(format = &quot;BED&quot;)</code> is a <code>GRanges</code>.</p>
<p>You can specify <code>genome</code> (for example <code>hg19</code>) and the function will try to make an effort to populated the <code>seqinfo</code> of the <code>GRanges</code>.</p>
<p>You can also use the <code>which</code> argument to selectively parse a subset of the file which overlaps a <code>GRanges</code>. This becomes much more efficient if the file has been tabix-indexed (see below).</p>
</div>
<div id="bigwig-files" class="section level2">
<h2>BigWig files</h2>
<p>BigWig files typically store whole-genome coverage vectors (or at least whole-genome data). For this reason, the R representation of a BigWig file is usually quite big, so it might be necessary to read it into R in small chunks.</p>
<p>As for BED files, <code>import(format=&quot;BigWig&quot;)</code> supports a <code>which</code> argument which is a <code>GRanges</code>. It output data type is a <code>GRanges</code> per default, but using the <code>as</code> agurment you can have <code>as=&quot;Rle&quot;</code> and a few other options.</p>
<p>The <code>import(format=&quot;BigWig&quot;)</code> does not support a <code>genome</code> argument.</p>
</div>
<div id="other-file-formats" class="section level2">
<h2>Other file formats</h2>
<ul>
<li>GFF</li>
<li>TwoBit</li>
<li>Wig</li>
<li>bedGRaph</li>
</ul>
</div>
<div id="extensive-example" class="section level2">
<h2>Extensive example</h2>
<p>Let us start an <code>AnnotationHub</code>:</p>
<pre class="r"><code>library(AnnotationHub)
ahub &lt;- AnnotationHub()
table(ahub$rdataclass)</code></pre>
<pre><code>## 
##       BigWigFile           biopax        ChainFile       data.frame 
##            10143                9             1113                6 
##    ExpressionSet           FaFile          GRanges    Inparanoid8Db 
##                1             5112            17365              268 
##            OrgDb SQLiteConnection       TwoBitFile          VcfFile 
##             1164                1              144               38</code></pre>
<p>At this point, you should have seen several of these file formats mentioned. The <code>GRanges</code> are usually directly constructed from BED file, and the <code>seqinfo</code> information is fully populated:</p>
<pre class="r"><code>ahub.gr &lt;- subset(ahub, rdataclass == &quot;GRanges&quot; &amp; species == &quot;Homo sapiens&quot;)
gr &lt;- ahub.gr[[1]]
gr</code></pre>
<pre><code>## GRanges object with 83523 ranges and 5 metadata columns:
##           seqnames                 ranges strand   |
##              &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt;   |
##       [1]    chr11 [ 65266509,  65266572]      +   |
##       [2]     chr1 [156675257, 156675415]      -   |
##       [3]    chr10 [ 33247091,  33247233]      -   |
##       [4]     chrM [    10469,     10473]      +   |
##       [5]    chr15 [ 69745106,  69745303]      +   |
##       ...      ...                    ...    ... ...
##   [83519]     chrX [100078869, 100078951]      +   |
##   [83520]     chrX [110535150, 110535159]      -   |
##   [83521]     chrX [118371391, 118371473]      +   |
##   [83522]     chrX [ 64254602,  64254777]      -   |
##   [83523]     chrX [ 76889467,  76889479]      -   |
##                                               name     score     level
##                                        &lt;character&gt; &lt;integer&gt; &lt;numeric&gt;
##       [1]  chr11:65266509:65266572:+:0.04:1.000000         0 20993.670
##       [2] chr1:156675257:156675415:-:0.35:1.000000         0 11580.471
##       [3]  chr10:33247091:33247233:-:0.32:1.000000         0  8098.093
##       [4]         chrM:10469:10473:+:4.32:0.000017         0  7538.406
##       [5]  chr15:69745106:69745303:+:1.62:1.000000         0  6673.187
##       ...                                      ...       ...       ...
##   [83519] chrX:100078869:100078951:+:2.69:0.003504         0    0.4826
##   [83520] chrX:110535150:110535159:-:3.77:0.000296         0    2.3895
##   [83521] chrX:118371391:118371473:+:0.28:0.011726         0    0.4826
##   [83522]   chrX:64254602:64254777:-:2.53:1.000000         0    0.5803
##   [83523]   chrX:76889467:76889479:-:1.26:0.000001         0    0.4826
##              signif    score2
##           &lt;numeric&gt; &lt;integer&gt;
##       [1]  3.36e-10         0
##       [2]  3.54e-10         0
##       [3]  3.82e-10         0
##       [4]  3.82e-10         0
##       [5]  4.09e-10         0
##       ...       ...       ...
##   [83519]         1         0
##   [83520]         1         0
##   [83521]         1         0
##   [83522]         1         0
##   [83523]         1         0
##   -------
##   seqinfo: 24 sequences from hg19 genome</code></pre>
<pre class="r"><code>seqinfo(gr)</code></pre>
<pre><code>## Seqinfo object with 24 sequences from hg19 genome:
##   seqnames seqlengths isCircular genome
##   chr1      249250621      FALSE   hg19
##   chr10     135534747      FALSE   hg19
##   chr11     135006516      FALSE   hg19
##   chr12     133851895      FALSE   hg19
##   chr13     115169878      FALSE   hg19
##   ...             ...        ...    ...
##   chr7      159138663      FALSE   hg19
##   chr8      146364022      FALSE   hg19
##   chr9      141213431      FALSE   hg19
##   chrM          16571      FALSE   hg19
##   chrX      155270560      FALSE   hg19</code></pre>
<p>Perhaps more interesting is the data in form of BigWig files.</p>
<pre class="r"><code>ahub.bw &lt;- subset(ahub, rdataclass == &quot;BigWigFile&quot; &amp; species == &quot;Homo sapiens&quot;)
ahub.bw</code></pre>
<pre><code>## AnnotationHub with 9828 records
## # snapshotDate(): 2015-08-17 
## # $dataprovider: BroadInstitute
## # $species: Homo sapiens
## # $rdataclass: BigWigFile
## # additional mcols(): taxonomyid, genome, description, tags,
## #   sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[[&quot;AH32002&quot;]]' 
## 
##             title                                  
##   AH32002 | E001-H3K4me1.fc.signal.bigwig          
##   AH32003 | E001-H3K4me3.fc.signal.bigwig          
##   AH32004 | E001-H3K9ac.fc.signal.bigwig           
##   AH32005 | E001-H3K9me3.fc.signal.bigwig          
##   AH32006 | E001-H3K27me3.fc.signal.bigwig         
##   ...       ...                                    
##   AH41825 | E125-H4K91ac.imputed.pval.signal.bigwig
##   AH41826 | E126-H4K91ac.imputed.pval.signal.bigwig
##   AH41827 | E127-H4K91ac.imputed.pval.signal.bigwig
##   AH41828 | E128-H4K91ac.imputed.pval.signal.bigwig
##   AH41829 | E129-H4K91ac.imputed.pval.signal.bigwig</code></pre>
<pre class="r"><code>bw &lt;- ahub.bw[[1]]
bw</code></pre>
<pre><code>## BigWigFile object
## resource: /Users/khansen/.AnnotationHub/37442</code></pre>
<p>This returns us a file name, ready for use by <code>import</code>.</p>
<pre class="r"><code>gr1 &lt;- gr[1:3]
out.gr &lt;- import(bw, which = gr1)
out.gr</code></pre>
<pre><code>## GRanges object with 14 ranges and 1 metadata column:
##        seqnames                 ranges strand   |             score
##           &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt;   |         &lt;numeric&gt;
##    [1]     chr1 [156675257, 156675262]      *   | 0.634060025215149
##    [2]     chr1 [156675263, 156675285]      *   | 0.501269996166229
##    [3]     chr1 [156675286, 156675297]      *   | 0.250600010156631
##    [4]     chr1 [156675298, 156675307]      *   | 0.315039992332458
##    [5]     chr1 [156675308, 156675332]      *   | 0.303939998149872
##    ...      ...                    ...    ... ...               ...
##   [10]     chr1 [156675392, 156675397]      *   | 0.303939998149872
##   [11]     chr1 [156675398, 156675401]      *   | 0.250600010156631
##   [12]     chr1 [156675402, 156675415]      *   |                 0
##   [13]    chr10 [ 33247091,  33247233]      *   |                 0
##   [14]    chr11 [ 65266509,  65266572]      *   |                 0
##   -------
##   seqinfo: 25 sequences from an unspecified genome</code></pre>
<p>This gives us the content in the form of a <code>GRanges</code>. Often, an <code>Rle</code> might be appropriate:</p>
<pre class="r"><code>out.rle &lt;- import(bw, which = gr1, as = &quot;Rle&quot;)
out.rle</code></pre>
<pre><code>## RleList of length 25
## $chr1
## numeric-Rle of length 249250621 with 13 runs
##   Lengths:         156675256                 6 ...          92575220
##   Values :                 0 0.634060025215149 ...                 0
## 
## $chr10
## numeric-Rle of length 135534747 with 1 run
##   Lengths: 135534747
##   Values :         0
## 
## $chr11
## numeric-Rle of length 135006516 with 1 run
##   Lengths: 135006516
##   Values :         0
## 
## $chr12
## numeric-Rle of length 133851895 with 1 run
##   Lengths: 133851895
##   Values :         0
## 
## $chr13
## numeric-Rle of length 115169878 with 1 run
##   Lengths: 115169878
##   Values :         0
## 
## ...
## &lt;20 more elements&gt;</code></pre>
<p>You can get all of <code>chr22</code> by</p>
<pre class="r"><code>gr.chr22 &lt;- GRanges(seqnames = &quot;chr22&quot;,
                    ranges = IRanges(start = 1, end = seqlengths(gr)[&quot;chr22&quot;]))
out.chr22 &lt;- import(bw, which = gr.chr22, as = &quot;Rle&quot;)
out.chr22[[&quot;chr22&quot;]]</code></pre>
<pre><code>## numeric-Rle of length 51304566 with 1381642 runs
##   Lengths:          16050196               194 ...             61301
##   Values :                 0 0.465829998254776 ...                 0</code></pre>
</div>
<div id="liftover" class="section level2">
<h2>LiftOver</h2>
<p>LiftOver is a popular tool from the UCSC Genome Browser for converting between different genome versions. The <em><a href="http://bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a></em> package also exposes this function through the <code>liftOver</code>. To use <code>liftOver</code> you need a so-called “chain” file describing how to convert from one genome to another. This can be obtained by hand from UCSC, or directly from <em><a href="http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a></em>.</p>
<p>We can re-use our <code>AnnotationHub</code>:</p>
<pre class="r"><code>ahub.chain &lt;- subset(ahub, rdataclass == &quot;ChainFile&quot; &amp; species == &quot;Homo sapiens&quot;)
query(ahub.chain, c(&quot;hg18&quot;, &quot;hg19&quot;))</code></pre>
<pre><code>## AnnotationHub with 2 records
## # snapshotDate(): 2015-08-17 
## # $dataprovider: UCSC
## # $species: Homo sapiens
## # $rdataclass: ChainFile
## # additional mcols(): taxonomyid, genome, description, tags,
## #   sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[[&quot;AH14149&quot;]]' 
## 
##             title                   
##   AH14149 | hg19ToHg18.over.chain.gz
##   AH14220 | hg18ToHg19.over.chain.gz</code></pre>
<pre class="r"><code>chain &lt;- ahub.chain[ahub.chain$title == &quot;hg19ToHg18.over.chain.gz&quot;]
chain &lt;- chain[[1]]
gr.hg18 &lt;- liftOver(gr, chain)
gr.hg18</code></pre>
<pre><code>## GRangesList object of length 83523:
## $1 
## GRanges object with 1 range and 5 metadata columns:
##       seqnames               ranges strand |
##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; |
##   [1]    chr11 [65023085, 65023148]      + |
##                                          name     score     level
##                                   &lt;character&gt; &lt;integer&gt; &lt;numeric&gt;
##   [1] chr11:65266509:65266572:+:0.04:1.000000         0  20993.67
##          signif    score2
##       &lt;numeric&gt; &lt;integer&gt;
##   [1]  3.36e-10         0
## 
## $2 
## GRanges object with 1 range and 5 metadata columns:
##       seqnames                 ranges strand |
##   [1]     chr1 [154941881, 154942039]      - |
##                                           name score    level   signif
##   [1] chr1:156675257:156675415:-:0.35:1.000000     0 11580.47 3.54e-10
##       score2
##   [1]      0
## 
## $3 
## GRanges object with 1 range and 5 metadata columns:
##       seqnames               ranges strand |
##   [1]    chr10 [33287097, 33287239]      - |
##                                          name score    level   signif
##   [1] chr10:33247091:33247233:-:0.32:1.000000     0 8098.093 3.82e-10
##       score2
##   [1]      0
## 
## ...
## &lt;83520 more elements&gt;
## -------
## seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
<p>This converts a <code>GRanges</code> into a <code>GRangesList</code>, why? This is because a single range (interval) may be split into multiple intervals in the other genome. So each element in the output correspond to a single range in the input. If the ranges are small, most ranges should be mapped to a single range. Let us look at the number of elements in output:</p>
<pre class="r"><code>table(elementLengths(gr.hg18))</code></pre>
<pre><code>## 
##     0     1     2     3 
##    47 83468     6     2</code></pre>
<p>Only a few ranges were not mapped and only a few were split.</p>
</div>
<div id="importing-directly-from-ucsc" class="section level2">
<h2>Importing directly from UCSC</h2>
<p>Using <em><a href="http://bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a></em> you can import tables and tracks directly from the UCSC Genome Browser. However, it is now possible to get a lot (all?) of this data from <em><a href="http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a></em> and this later package seems friendlier.</p>
<p>It is possible that not all tracks / tables and/or all information from the different track / tables from UCSC are exposed in <em><a href="http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a></em>.</p>
<p>See a detailed exposition in the package vignette.</p>
</div>
<div id="tabix-indexing" class="section level2">
<h2>Tabix indexing</h2>
<p>Tabix indexing is a way to index a text file with chromosomal positions for random access. This will greatly speed up any querying of such a file. The <a href="http://www.htslib.org/doc/tabix.html">tabix</a> functionality was introduced in the SAMtools library; this library was later renamed to htslib.</p>
<p>Tabix indexing is usually something you do at the command line, but there is also the convenient possibility of doing it from inside Bioconductor using <code>indexTabix</code> from the <em><a href="http://bioconductor.org/packages/release/bioc/html/Rsamtools.html">Rsamtools</a></em> package. First however, the file needs to be bgzip2 compressed, which you can do using the <code>bgzip2</code> function. A full pipeline, using an example SAM file from <em><a href="http://bioconductor.org/packages/release/bioc/html/Rsamtools.html">Rsamtools</a></em> is</p>
<pre class="r"><code>library(Rsamtools)
from &lt;- system.file(&quot;extdata&quot;, &quot;ex1.sam&quot;, package=&quot;Rsamtools&quot;,
                    mustWork=TRUE)
from</code></pre>
<pre><code>## [1] &quot;/Library/Frameworks/R.framework/Versions/3.2/Resources/library/Rsamtools/extdata/ex1.sam&quot;</code></pre>
<pre class="r"><code>to &lt;- tempfile()
zipped &lt;- bgzip(from, to)
idx &lt;- indexTabix(zipped, &quot;sam&quot;)</code></pre>
<p>see also the help page for <code>indexTabix</code>.</p>
</div>
<div id="sessioninfo" class="section level2">
<h2>SessionInfo</h2>
<p></p>
<pre><code>## R version 3.2.1 (2015-06-18)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: OS X 10.10.5 (Yosemite)
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] Rsamtools_1.20.4     Biostrings_2.36.3    XVector_0.8.0       
##  [4] AnnotationHub_2.0.3  rtracklayer_1.28.8   GenomicRanges_1.20.5
##  [7] GenomeInfoDb_1.4.2   IRanges_2.2.7        S4Vectors_0.6.3     
## [10] BiocGenerics_0.14.0  BiocStyle_1.6.0      rmarkdown_0.7       
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.0                  BiocInstaller_1.18.4        
##  [3] formatR_1.2                  futile.logger_1.4.1         
##  [5] bitops_1.0-6                 futile.options_1.0.0        
##  [7] tools_3.2.1                  zlibbioc_1.14.0             
##  [9] digest_0.6.8                 evaluate_0.7.2              
## [11] RSQLite_1.0.0                shiny_0.12.2                
## [13] DBI_0.3.1                    curl_0.9.2                  
## [15] yaml_2.1.13                  stringr_1.0.0               
## [17] httr_1.0.0                   knitr_1.11                  
## [19] Biobase_2.28.0               R6_2.1.0                    
## [21] AnnotationDbi_1.30.1         XML_3.98-1.3                
## [23] BiocParallel_1.2.20          lambda.r_1.1.7              
## [25] magrittr_1.5                 htmltools_0.2.6             
## [27] GenomicAlignments_1.4.1      mime_0.3                    
## [29] interactiveDisplayBase_1.6.0 xtable_1.7-4                
## [31] httpuv_1.3.3                 stringi_0.5-5               
## [33] RCurl_1.95-4.7</code></pre>
<div class="references">

</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
