<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Kasper D. Hansen" />


<title>Count Based RNA-seq analysis</title>


<style type="text/css">code{white-space: pre;}</style>
<link href="data:text/css;charset=utf-8,pre%20%2Eoperator%2C%0Apre%20%2Eparen%20%7B%0Acolor%3A%20rgb%28104%2C%20118%2C%20135%29%0A%7D%0Apre%20%2Eliteral%20%7B%0Acolor%3A%20%23990073%0A%7D%0Apre%20%2Enumber%20%7B%0Acolor%3A%20%23099%3B%0A%7D%0Apre%20%2Ecomment%20%7B%0Acolor%3A%20%23998%3B%0Afont%2Dstyle%3A%20italic%0A%7D%0Apre%20%2Ekeyword%20%7B%0Acolor%3A%20%23900%3B%0Afont%2Dweight%3A%20bold%0A%7D%0Apre%20%2Eidentifier%20%7B%0Acolor%3A%20rgb%280%2C%200%2C%200%29%3B%0A%7D%0Apre%20%2Estring%20%7B%0Acolor%3A%20%23d14%3B%0A%7D%0A" rel="stylesheet" type="text/css" />
<script src="data:application/x-javascript;base64,
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();

"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>


<link href="data:text/css;charset=utf-8,body%2C%20td%20%7B%0Afont%2Dfamily%3A%20sans%2Dserif%3B%0Abackground%2Dcolor%3A%20white%3B%0Afont%2Dsize%3A%2013px%3B%0A%7D%0Abody%20%7B%0Amax%2Dwidth%3A%20800px%3B%0Amargin%3A%200%20auto%3B%0Apadding%3A%201em%201em%202em%3B%0Aline%2Dheight%3A%2020px%3B%0A%7D%0A%0Ap%2C%20pre%20%7B%20margin%3A%200em%200em%201em%3B%0A%7D%0A%0Aimg%2C%20table%20%7B%0Amargin%3A%200em%20auto%201em%3B%0A%7D%0Ap%20%7B%0Atext%2Dalign%3A%20justify%3B%0A%7D%0Att%2C%20code%2C%20pre%20%7B%0Afont%2Dfamily%3A%20%27DejaVu%20Sans%20Mono%27%2C%20%27Droid%20Sans%20Mono%27%2C%20%27Lucida%20Console%27%2C%20Consolas%2C%20Monaco%2C%20monospace%3B%0A%7D%0Ah1%2C%20h2%2C%20h3%2C%20h4%2C%20h5%2C%20h6%20%7B%20font%2Dfamily%3A%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Amargin%3A%201%2E2em%200em%200%2E6em%200em%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%20250%25%3B%0Afont%2Dweight%3A%20normal%3B%0Acolor%3A%20%2387b13f%3B%0Aline%2Dheight%3A%201%2E1em%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%20160%25%3B%0Afont%2Dweight%3A%20normal%3B%0Aline%2Dheight%3A%201%2E4em%3B%0Aborder%2Dbottom%3A%201px%20%231a81c2%20solid%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%20130%25%3B%20%7D%0Ah2%2C%20h3%20%7B%0Acolor%3A%20%231a81c2%3B%0A%7D%0Ah4%2C%20h5%2C%20h6%20%7B%0Afont%2Dsize%3A115%25%3B%0A%7D%20%0A%0Aa%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Aactive%20%7B%20outline%3A%20none%3B%20%7D%0Aa%3Avisited%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Ahover%20%7B%20color%3A%20%234c94c2%3B%20%7D%0Apre%2C%20img%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Adisplay%3A%20block%3B%0A%7D%0Apre%20%7B%0Aborder%3A%200px%20none%3B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0Awhite%2Dspace%3A%20pre%3B%0Aoverflow%2Dx%3A%20auto%3B%0A%7D%0Apre%20code%20%7B%0Aborder%3A%201px%20%23aaa%20dashed%3B%0Abackground%2Dcolor%3A%20white%3B%0Adisplay%3A%20block%3B%0Apadding%3A%201em%3B%20color%3A%20%23111%3B%0Aoverflow%2Dx%3A%20inherit%3B%0A%7D%0A%0Apre%20code%5Bclass%5D%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Apre%5Bclass%5D%20code%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Acode%20%7B%20background%2Dcolor%3A%20transparent%3B%0Acolor%3A%20%2387b13f%3B%0Afont%2Dsize%3A%2092%25%3B%0A%7D%0A%0Atable%2C%20td%2C%20th%20%7B%0Aborder%3A%20none%3B%0Apadding%3A%200%200%2E5em%3B%0A%7D%0A%0Atbody%20tr%3Anth%2Dchild%28odd%29%20td%20%7B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0A%7D%0Ablockquote%20%7B%0Acolor%3A%23666666%3B%0Amargin%3A0%3B%0Apadding%2Dleft%3A%201em%3B%0Aborder%2Dleft%3A%200%2E5em%20%23EEE%20solid%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200px%3B%0Aborder%2Dbottom%3A%20none%3B%0Aborder%2Dtop%2Dwidth%3A%20thin%3B%0Aborder%2Dtop%2Dstyle%3A%20dotted%3B%0Aborder%2Dtop%2Dcolor%3A%20%23999999%3B%0A%7D%0A%40media%20print%20%7B%0A%2A%20%7B%0Abackground%3A%20transparent%20%21important%3B%0Acolor%3A%20black%20%21important%3B%0Afilter%3Anone%20%21important%3B%0A%2Dms%2Dfilter%3A%20none%20%21important%3B%0A%7D%0Abody%20%7B%0Afont%2Dsize%3A12pt%3B%0Amax%2Dwidth%3A100%25%3B%0A%7D%0Aa%2C%20a%3Avisited%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Avisibility%3A%20hidden%3B%0Apage%2Dbreak%2Dbefore%3A%20always%3B%0A%7D%0Apre%2C%20blockquote%20%7B%0Apadding%2Dright%3A%201em%3B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Atr%2C%20img%20%7B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%20%21important%3B%0A%7D%0A%40page%20%3Aleft%20%7B%0Amargin%3A%2015mm%2020mm%2015mm%2010mm%3B%0A%7D%0A%40page%20%3Aright%20%7B%0Amargin%3A%2015mm%2010mm%2015mm%2020mm%3B%0A%7D%0Ap%2C%20h2%2C%20h3%20%7B%0Aorphans%3A%203%3B%20widows%3A%203%3B%0A%7D%0Ah2%2C%20h3%20%7B%0Apage%2Dbreak%2Dafter%3A%20avoid%3B%0A%7D%0A%7D%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title">Count Based RNA-seq analysis</h1>
<h4 class="author"><em>Kasper D. Hansen</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#dependencies">Dependencies</a></li>
<li><a href="#overview">Overview</a></li>
<li><a href="#other-resources">Other Resources</a></li>
<li><a href="#rna-seq-count-data">RNA-seq count data</a></li>
<li><a href="#statistical-issues">Statistical issues</a></li>
<li><a href="#the-data">The Data</a></li>
<li><a href="#edger">edgeR</a></li>
<li><a href="#deseq2">DESeq2</a></li>
<li><a href="#sessioninfo">SessionInfo</a></li>
</ul>
</div>

<div id="dependencies" class="section level2">
<h2>Dependencies</h2>
<p>This document has the following dependencies:</p>
<pre class="r"><code>library(DESeq2)
library(edgeR)</code></pre>
<p>Use the following commands to install these packages in R.</p>
<pre class="r"><code>source(&quot;http://www.bioconductor.org/biocLite.R&quot;)
biocLite(c(&quot;DESeq2&quot;, &quot;edgeR))</code></pre>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>RNA seq data is often analyzed by creating a count matrix of gene counts per sample. This matrix is analyzed using count-based models, often built on the negative binomial distribution. Popular packages for this includes <em><a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a></em> and <em><a href="http://bioconductor.org/packages/release/bioc/html/DESeq.html">DESeq</a></em> / <em><a href="http://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a></em>.</p>
<p>This type of analysis discards part of the information in the RNA sequencing reads, but we have a good understanding of how to analyze this type of data.</p>
</div>
<div id="other-resources" class="section level2">
<h2>Other Resources</h2>
<ul>
<li>The vignette from the <a href="http://bioconductor.org/packages/edgeR">edgeR webpage</a>.</li>
<li>The vignette from the <a href="http://bioconductor.org/packages/DESeq2">DESeq2 webpage</a>.</li>
<li>The <a href="http://bioconductor.org/help/workflows/rnaseqGene/">RNA-seq workflow</a>.</li>
</ul>
</div>
<div id="rna-seq-count-data" class="section level2">
<h2>RNA-seq count data</h2>
<p>One simple way of analyzing RNA sequencing data is to make it look like microarray data. This is done by counting how many reads in each sample overlaps a gene. There are many ways to do this. It obviously depends on the annotation used, but also on how it is decided that a read overlaps a region. Of specific concern is which genomic regions are part of a gene with multiple transcripts.</p>
<p>There are no consensus on this process and the different choices one make is known to affect the outcome.</p>
<p>Tools for doing gene counting includes</p>
<ul>
<li>by using <code>featureCounts()</code> from the <em><a href="http://bioconductor.org/packages/release/bioc/html/Rsubread.html">Rsubread</a></em> package.</li>
<li>the <a href="http://www-huber.embl.de/users/anders/HTSeq/">HTSeq</a> package (this is a python package, not a Bioconductor package).</li>
<li>by using <code>summarizeOverlaps()</code> from the <em><a href="http://bioconductor.org/packages/release/bioc/html/GenomicAlignments.html">GenomicAlignments</a></em> package.</li>
</ul>
<p>and there are other alternatives. Many people seem to write their own counting pipeline.</p>
<p>Reducing RNA sequencing data to a single integer per gene is obvious a simplification. Indeed it ignores some of the main reasons for doing RNA sequencing, including assessing alternative splicing. On the other hand, we understand the statistical properties of this procedure well, and it delivers a basic insight into something that most researcher wants to know. Finally, this approach requires the different genomic regions to be known beforehand.</p>
</div>
<div id="statistical-issues" class="section level2">
<h2>Statistical issues</h2>
<p>In RNA-seq data analysis we often see that many genes (up to 50%) have little or no expression. It is common to pre-filter (remove) these genes prior to analysis. In general genomics filtering might be beneficial to your analysis, but this discussion is outside the scope of this document.</p>
<p><strong>Note:</strong> The analysis presented below is <strong>extremely</strong> superficial. Consider this a very quick introduction to the workflow of these two packages.</p>
</div>
<div id="the-data" class="section level2">
<h2>The Data</h2>
<p>We will be using the <em><a href="http://bioconductor.org/packages/release/data/experiment/html/airway.html">airway</a></em> dataset which contains RNA-seq data in the form of a <code>SummarizedExperiment</code>. Lets load the data and have a look</p>
<pre class="r"><code>library(airway)
data(airway)
airway</code></pre>
<pre><code>## class: SummarizedExperiment 
## dim: 64102 8 
## exptData(1): ''
## assays(1): counts
## rownames(64102): ENSG00000000003 ENSG00000000005 ... LRG_98 LRG_99
## rowRanges metadata column names(0):
## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
## colData names(9): SampleName cell ... Sample BioSample</code></pre>
<pre class="r"><code>assay(airway, &quot;counts&quot;)[1:3, 1:3]</code></pre>
<pre><code>##                 SRR1039508 SRR1039509 SRR1039512
## ENSG00000000003        679        448        873
## ENSG00000000005          0          0          0
## ENSG00000000419        467        515        621</code></pre>
<pre class="r"><code>airway$dex</code></pre>
<pre><code>## [1] untrt trt   untrt trt   untrt trt   untrt trt  
## Levels: trt untrt</code></pre>
<p>The main variable of interest is <code>dex</code> which takes on levels <code>trt</code> (treated) and <code>untrt</code> (untreated). The first level will be the reference level for this factor, so we use <code>relevel()</code> to set the <code>untrt</code> level as reference; this is much easier to interpret.</p>
<pre class="r"><code>airway$dex &lt;- relevel(airway$dex, &quot;untrt&quot;)
airway$dex</code></pre>
<pre><code>## [1] untrt trt   untrt trt   untrt trt   untrt trt  
## Levels: untrt trt</code></pre>
<p>There is rich information about which gene model was used for each gene:</p>
<pre class="r"><code>granges(airway)</code></pre>
<pre><code>## GRangesList object of length 64102:
## $ENSG00000000003 
## GRanges object with 17 ranges and 2 metadata columns:
##        seqnames               ranges strand   |   exon_id       exon_name
##           &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt;   | &lt;integer&gt;     &lt;character&gt;
##    [1]        X [99883667, 99884983]      -   |    667145 ENSE00001459322
##    [2]        X [99885756, 99885863]      -   |    667146 ENSE00000868868
##    [3]        X [99887482, 99887565]      -   |    667147 ENSE00000401072
##    [4]        X [99887538, 99887565]      -   |    667148 ENSE00001849132
##    [5]        X [99888402, 99888536]      -   |    667149 ENSE00003554016
##    ...      ...                  ...    ... ...       ...             ...
##   [13]        X [99890555, 99890743]      -   |    667156 ENSE00003512331
##   [14]        X [99891188, 99891686]      -   |    667158 ENSE00001886883
##   [15]        X [99891605, 99891803]      -   |    667159 ENSE00001855382
##   [16]        X [99891790, 99892101]      -   |    667160 ENSE00001863395
##   [17]        X [99894942, 99894988]      -   |    667161 ENSE00001828996
## 
## ...
## &lt;64101 more elements&gt;
## -------
## seqinfo: 722 sequences (1 circular) from an unspecified genome</code></pre>
</div>
<div id="edger" class="section level2">
<h2>edgeR</h2>
<p>The <em><a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a></em> is very similar in terms of data structures and functionality to the <em><a href="http://bioconductor.org/packages/release/bioc/html/limma.html">limma</a></em>. Whereas <em><a href="http://bioconductor.org/packages/release/bioc/html/limma.html">limma</a></em> allows us to operate directy on <code>ExpressionSet</code>s, edgeR does not work directly with <code>SummarizedExperiment</code>. We first need to put out data into an edgeR specific container.</p>
<pre class="r"><code>library(edgeR)
dge &lt;- DGEList(counts = assay(airway, &quot;counts&quot;),
               group = airway$dex)
dge$samples &lt;- merge(dge$samples,
                     as.data.frame(colData(airway)),
                     by = 0)
dge$genes &lt;- data.frame(name = names(rowRanges(airway)),
                        stringsAsFactors = FALSE)</code></pre>
<p>This object has something called the <code>group</code> which is the basic experimental group for each sample. It also has <code>$samples</code> (the pheno data) which - weirdly - cannot be set when you create the <code>DGEList</code> object, so we set it afterwards. The <code>$genes</code> is a <code>data.frame</code> so we cannot include the rich gene model information we had in the <code>SummarizedExperiment</code>.</p>
<p>Having set up the input object, we now proceed as follows.</p>
<p>First we estimate the normalization factors or effective library sizes</p>
<pre class="r"><code>dge &lt;- calcNormFactors(dge)</code></pre>
<p>Next we estimate the dispersion (variance). There are multiple ways to do this, and the weird two-step procedure is necessary.</p>
<pre class="r"><code>dge &lt;- estimateGLMCommonDisp(dge)
dge &lt;- estimateGLMTagwiseDisp(dge)</code></pre>
<p>Now we setup the design matrix and do a <code>glmFit()</code>, similar to <em><a href="http://bioconductor.org/packages/release/bioc/html/limma.html">limma</a></em></p>
<pre class="r"><code>design &lt;- model.matrix(~dge$samples$group)
fit &lt;- glmFit(dge, design)</code></pre>
<p>Now it is time to do a test and extract the top hits</p>
<pre class="r"><code>lrt &lt;- glmLRT(fit, coef = 2)
topTags(lrt)</code></pre>
<pre><code>## Coefficient:  dge$samples$grouptrt 
##                  name     logFC    logCPM       LR       PValue       FDR
## 56702 ENSG00000265702  5.793444 -2.057490 18.86584 1.402419e-05 0.3032692
## 26235 ENSG00000222939  5.279691 -2.393440 18.47944 1.717474e-05 0.3032692
## 14922 ENSG00000179593 10.102066  1.658949 18.01816 2.188075e-05 0.3032692
## 55622 ENSG00000264569  5.520191 -2.235712 17.90243 2.325244e-05 0.3032692
## 6667  ENSG00000132518  6.536657 -1.591221 17.60321 2.721293e-05 0.3032692
## 28978 ENSG00000226453 -4.891970 -2.504215 16.30160 5.401843e-05 0.3032692
## 3751  ENSG00000109906  7.082997  4.162779 16.19655 5.709795e-05 0.3032692
## 6058  ENSG00000128285 -5.209306 -1.287458 16.16211 5.814574e-05 0.3032692
## 38438 ENSG00000237949  3.828161 -2.279109 15.98393 6.388248e-05 0.3032692
## 363   ENSG00000015413  4.764661 -2.606451 15.93160 6.567293e-05 0.3032692</code></pre>
</div>
<div id="deseq2" class="section level2">
<h2>DESeq2</h2>
<p>Like <em><a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a></em>, DESeq2 requires us to put the data into a package-specific container (a <code>DESeqDataSet</code>). But unlike edgeR, it is pretty easy.</p>
<pre class="r"><code>library(DESeq2)
dds &lt;- DESeqDataSet(airway, design = ~ dex)</code></pre>
<p>Note that the design of the experiment is stored inside the object. The last variable (in case multiple variables are list) will be the variable of interest which is report in the different results outputs.</p>
<p>Fitting the model is simple</p>
<pre class="r"><code>dds &lt;- DESeq(dds)</code></pre>
<p>and then all we need to do is get the results. Note that the results are not ordered, so we do that.</p>
<pre class="r"><code>res &lt;- results(dds)
res &lt;- res[order(res$padj),]
res</code></pre>
<pre><code>## log2 fold change (MAP): dex trt vs untrt 
## Wald test p-value: dex trt vs untrt 
## DataFrame with 64102 rows and 6 columns
##                   baseMean log2FoldChange      lfcSE      stat
##                  &lt;numeric&gt;      &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;
## ENSG00000152583   997.4398       4.285847 0.19605831  21.86006
## ENSG00000148175 11193.7188       1.434388 0.08411178  17.05336
## ENSG00000179094   776.5967       2.984244 0.18864120  15.81968
## ENSG00000109906   385.0710       5.137007 0.33077733  15.53010
## ENSG00000134686  2737.9820       1.368176 0.09059205  15.10261
## ...                    ...            ...        ...       ...
## LRG_94                   0             NA         NA        NA
## LRG_96                   0             NA         NA        NA
## LRG_97                   0             NA         NA        NA
## LRG_98                   0             NA         NA        NA
## LRG_99                   0             NA         NA        NA
##                        pvalue          padj
##                     &lt;numeric&gt;     &lt;numeric&gt;
## ENSG00000152583 6.235873e-106 1.089469e-101
## ENSG00000148175  3.300017e-65  2.882730e-61
## ENSG00000179094  2.276377e-56  1.325686e-52
## ENSG00000109906  2.170243e-54  9.479080e-51
## ENSG00000134686  1.556576e-51  5.438987e-48
## ...                       ...           ...
## LRG_94                     NA            NA
## LRG_96                     NA            NA
## LRG_97                     NA            NA
## LRG_98                     NA            NA
## LRG_99                     NA            NA</code></pre>
</div>
<div id="sessioninfo" class="section level2">
<h2>SessionInfo</h2>
<p></p>
<pre><code>## R version 3.2.1 (2015-06-18)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: OS X 10.10.5 (Yosemite)
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] airway_0.102.0            edgeR_3.10.2             
##  [3] limma_3.24.15             DESeq2_1.8.1             
##  [5] RcppArmadillo_0.5.500.2.0 Rcpp_0.12.0              
##  [7] GenomicRanges_1.20.6      GenomeInfoDb_1.4.2       
##  [9] IRanges_2.2.7             S4Vectors_0.6.5          
## [11] BiocGenerics_0.14.0       BiocStyle_1.6.0          
## [13] rmarkdown_0.8            
## 
## loaded via a namespace (and not attached):
##  [1] RColorBrewer_1.1-2   formatR_1.2          futile.logger_1.4.1 
##  [4] plyr_1.8.3           XVector_0.8.0        futile.options_1.0.0
##  [7] tools_3.2.1          rpart_4.1-10         digest_0.6.8        
## [10] RSQLite_1.0.0        annotate_1.46.1      evaluate_0.7.2      
## [13] gtable_0.1.2         lattice_0.20-33      DBI_0.3.1           
## [16] yaml_2.1.13          proto_0.3-10         gridExtra_2.0.0     
## [19] genefilter_1.50.0    cluster_2.0.3        stringr_1.0.0       
## [22] knitr_1.11           locfit_1.5-9.1       nnet_7.3-10         
## [25] grid_3.2.1           Biobase_2.28.0       AnnotationDbi_1.30.1
## [28] XML_3.98-1.3         survival_2.38-3      BiocParallel_1.2.20 
## [31] foreign_0.8-65       latticeExtra_0.6-26  Formula_1.2-1       
## [34] geneplotter_1.46.0   ggplot2_1.0.1        reshape2_1.4.1      
## [37] lambda.r_1.1.7       magrittr_1.5         scales_0.3.0        
## [40] Hmisc_3.16-0         htmltools_0.2.6      MASS_7.3-43         
## [43] splines_3.2.1        xtable_1.7-4         colorspace_1.2-6    
## [46] stringi_0.5-5        acepack_1.3-3.3      munsell_0.4.2</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
