<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Kasper D. Hansen" />


<title>Analysis of 450k DNA methylation data with minfi</title>


<style type="text/css">code{white-space: pre;}</style>
<link href="data:text/css;charset=utf-8,pre%20%2Eoperator%2C%0Apre%20%2Eparen%20%7B%0Acolor%3A%20rgb%28104%2C%20118%2C%20135%29%0A%7D%0Apre%20%2Eliteral%20%7B%0Acolor%3A%20%23990073%0A%7D%0Apre%20%2Enumber%20%7B%0Acolor%3A%20%23099%3B%0A%7D%0Apre%20%2Ecomment%20%7B%0Acolor%3A%20%23998%3B%0Afont%2Dstyle%3A%20italic%0A%7D%0Apre%20%2Ekeyword%20%7B%0Acolor%3A%20%23900%3B%0Afont%2Dweight%3A%20bold%0A%7D%0Apre%20%2Eidentifier%20%7B%0Acolor%3A%20rgb%280%2C%200%2C%200%29%3B%0A%7D%0Apre%20%2Estring%20%7B%0Acolor%3A%20%23d14%3B%0A%7D%0A" rel="stylesheet" type="text/css" />
<script src="data:application/x-javascript;base64,
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();

"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>


<link href="data:text/css;charset=utf-8,body%2C%20td%20%7B%0Afont%2Dfamily%3A%20sans%2Dserif%3B%0Abackground%2Dcolor%3A%20white%3B%0Afont%2Dsize%3A%2013px%3B%0A%7D%0Abody%20%7B%0Amax%2Dwidth%3A%20800px%3B%0Amargin%3A%200%20auto%3B%0Apadding%3A%201em%201em%202em%3B%0Aline%2Dheight%3A%2020px%3B%0A%7D%0A%0Adiv%23TOC%20li%20%7B%0Alist%2Dstyle%3Anone%3B%0Abackground%2Dimage%3Anone%3B%0Abackground%2Drepeat%3Anone%3B%0Abackground%2Dposition%3A0%3B%0A%7D%0A%0Ap%2C%20pre%20%7B%20margin%3A%200em%200em%201em%3B%0A%7D%0A%0Aimg%2C%20table%20%7B%0Amargin%3A%200em%20auto%201em%3B%0A%7D%0Ap%20%7B%0Atext%2Dalign%3A%20justify%3B%0A%7D%0Att%2C%20code%2C%20pre%20%7B%0Afont%2Dfamily%3A%20%27DejaVu%20Sans%20Mono%27%2C%20%27Droid%20Sans%20Mono%27%2C%20%27Lucida%20Console%27%2C%20Consolas%2C%20Monaco%2C%20monospace%3B%0A%7D%0Ah1%2C%20h2%2C%20h3%2C%20h4%2C%20h5%2C%20h6%20%7B%20font%2Dfamily%3A%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Amargin%3A%201%2E2em%200em%200%2E6em%200em%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%2Etitle%20%7B%0Afont%2Dsize%3A%20250%25%3B%0Afont%2Dweight%3A%20normal%3B%0Acolor%3A%20%2387b13f%3B%0Aline%2Dheight%3A%201%2E1em%3B%0Amargin%2Dtop%3A%200px%3B%0Aborder%2Dbottom%3A%200px%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%20160%25%3B%0Afont%2Dweight%3A%20normal%3B%0Aline%2Dheight%3A%201%2E4em%3B%0Aborder%2Dbottom%3A%201px%20%231a81c2%20solid%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%20130%25%3B%20%7D%0Ah1%2C%20h2%2C%20h3%20%7B%0Acolor%3A%20%231a81c2%3B%0A%7D%0Ah3%2C%20h4%2C%20h5%2C%20h6%20%7B%0Afont%2Dsize%3A115%25%3B%0A%7D%20%0A%0Aa%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Aactive%20%7B%20outline%3A%20none%3B%20%7D%0Aa%3Avisited%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Ahover%20%7B%20color%3A%20%234c94c2%3B%20%7D%0Apre%2C%20img%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Adisplay%3A%20block%3B%0A%7D%0Apre%20%7B%0Aborder%3A%200px%20none%3B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0Awhite%2Dspace%3A%20pre%3B%0Aoverflow%2Dx%3A%20auto%3B%0A%7D%0Apre%20code%20%7B%0Aborder%3A%201px%20%23aaa%20dashed%3B%0Abackground%2Dcolor%3A%20white%3B%0Adisplay%3A%20block%3B%0Apadding%3A%201em%3B%20color%3A%20%23111%3B%0Aoverflow%2Dx%3A%20inherit%3B%0A%7D%0A%0Apre%20code%5Bclass%5D%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Apre%5Bclass%5D%20code%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Acode%20%7B%20background%2Dcolor%3A%20transparent%3B%0Acolor%3A%20%2387b13f%3B%0Afont%2Dsize%3A%2092%25%3B%0A%7D%0A%0Atable%2C%20td%2C%20th%20%7B%0Aborder%3A%20none%3B%0Apadding%3A%200%200%2E5em%3B%0A%7D%0A%0Atbody%20tr%3Anth%2Dchild%28odd%29%20td%20%7B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0A%7D%0Ablockquote%20%7B%0Acolor%3A%23666666%3B%0Amargin%3A0%3B%0Apadding%2Dleft%3A%201em%3B%0Aborder%2Dleft%3A%200%2E5em%20%23EEE%20solid%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200px%3B%0Aborder%2Dbottom%3A%20none%3B%0Aborder%2Dtop%2Dwidth%3A%20thin%3B%0Aborder%2Dtop%2Dstyle%3A%20dotted%3B%0Aborder%2Dtop%2Dcolor%3A%20%23999999%3B%0A%7D%0Aspan%2Eheader%2Dsection%2Dnumber%20%7B%0Apadding%2Dright%3A%201em%3B%0A%7D%0Aspan%2Etoc%2Dsection%2Dnumber%3A%3Aafter%20%7B%0Acontent%3A%20%22%20%22%3B%0Awhite%2Dspace%3A%20pre%3B%0A%7D%0A%40media%20print%20%7B%0A%2A%20%7B%0Abackground%3A%20transparent%20%21important%3B%0Acolor%3A%20black%20%21important%3B%0Afilter%3Anone%20%21important%3B%0A%2Dms%2Dfilter%3A%20none%20%21important%3B%0A%7D%0Abody%20%7B%0Afont%2Dsize%3A12pt%3B%0Amax%2Dwidth%3A100%25%3B%0A%7D%0Aa%2C%20a%3Avisited%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Avisibility%3A%20hidden%3B%0Apage%2Dbreak%2Dbefore%3A%20always%3B%0A%7D%0Apre%2C%20blockquote%20%7B%0Apadding%2Dright%3A%201em%3B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Atr%2C%20img%20%7B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%20%21important%3B%0A%7D%0A%40page%20%3Aleft%20%7B%0Amargin%3A%2015mm%2020mm%2015mm%2010mm%3B%0A%7D%0A%40page%20%3Aright%20%7B%0Amargin%3A%2015mm%2010mm%2015mm%2020mm%3B%0A%7D%0Ap%2C%20h2%2C%20h3%20%7B%0Aorphans%3A%203%3B%20widows%3A%203%3B%0A%7D%0Ah2%2C%20h3%20%7B%0Apage%2Dbreak%2Dafter%3A%20avoid%3B%0A%7D%0A%7D%0A" rel="stylesheet" type="text/css" />

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if(links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>

</head>

<body>


<div id="header">
<h1 class="title">Analysis of 450k DNA methylation data with minfi</h1>
<h4 class="author"><em>Kasper D. Hansen</em></h4>
</div>

<h1>Contents</h1>
<div id="TOC">
<ul>
<li><a href="#dependencies">Dependencies</a></li>
<li><a href="#overview">Overview</a></li>
<li><a href="#dna-methylation">DNA methylation</a></li>
<li><a href="#array-design">Array Design</a></li>
<li><a href="#data">Data</a></li>
<li><a href="#preprocessing">Preprocessing</a></li>
<li><a href="#differential-methylation">Differential Methylation</a></li>
<li><a href="#other-resources">Other Resources</a></li>
<li><a href="#sessioninfo">SessionInfo</a></li>
<li><a href="#corrections">Corrections</a></li>
</ul>
</div>

<div id="dependencies" class="section level2">
<h2>Dependencies</h2>
<p>This document has the following dependencies:</p>
<pre class="r"><code>library(minfi)
library(GEOquery)</code></pre>
<p>Use the following commands to install these packages in R.</p>
<pre class="r"><code>source(&quot;http://www.bioconductor.org/biocLite.R&quot;)
biocLite(c(&quot;minfi&quot;, &quot;GEOquery&quot;))</code></pre>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>The Illumina 450k DNA methylation microarray is a cheap and relatively comprehensive array for assaying DNA methylation. It is the platform of choice for large sample profiling of DNA methylation, especially for so-called EWAS (Epigenome-wide association studies). The studies are like GWAS (genome-wide association studies) but instead of associated a phenotype (like disease) with genotype (typically measured on SNP arrays or exome/whole-genome sequencing) they associated phenotype with epigenotype.</p>
</div>
<div id="dna-methylation" class="section level2">
<h2>DNA methylation</h2>
<p>DNA methylation is a chemical modification of DNA. In humans, it occurs at CpG dinucleotides where the ‘C’ can be methylated or not. The methylation state of a given locus in a single cell is binary (technically tertiary since we have two copies of most chromosomes) but we measure DNA methylation across a population of cells. We observe that some loci has intermediate methylation values (between 0 and 1) and we use the methylation percentage (or Beta-value) to describe this.</p>
<p>The goal of most analyses of DNA methylation is to associate changes in phenotype with changes in methylation in one or more loci.</p>
</div>
<div id="array-design" class="section level2">
<h2>Array Design</h2>
<p>The 450k array has a very unusual design, which to some extent impact analysis. It is really a mixture of a two-color array and two one-color arrays. There are two main types of probes (type I and type II) and the probe design affects the signal distribution of the probe.</p>
<p>The raw data format for the 450k array is known as IDAT. Because the array is measured in two different colors, there are two files for each sample, typically with the extention <code>_Grn.idat</code> and <code>_Red.idat</code>. Illumina’s software suite for analysis of this array is called GenomeStudio. It is not unusual for practitioners to only have access to processed data from GenomeStudio instead of the raw IDAT files, but I and others have shown that there is information in the IDAT files which are beneficial to analysis.</p>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p>We will access a dataset created with the intention of studying acute mania. Serum samples were obtained from individuals hospitalized with acute mania as well as unaffected controls.</p>
<p>We want to obtain the IDAT files which are available as supplementary data. Far from all 450k datasets on GEO has IDAT files available. First we download the files.</p>
<pre class="r"><code>library(GEOquery)
getGEOSuppFiles(&quot;GSE68777&quot;)
untar(&quot;GSE68777/GSE68777_RAW.tar&quot;, exdir = &quot;GSE68777/idat&quot;)
head(list.files(&quot;GSE68777/idat&quot;, pattern = &quot;idat&quot;))</code></pre>
<pre><code>## [1] &quot;GSM1681154_5958091019_R03C02_Grn.idat&quot;   
## [2] &quot;GSM1681154_5958091019_R03C02_Grn.idat.gz&quot;
## [3] &quot;GSM1681154_5958091019_R03C02_Red.idat&quot;   
## [4] &quot;GSM1681154_5958091019_R03C02_Red.idat.gz&quot;
## [5] &quot;GSM1681155_5935446005_R05C01_Grn.idat&quot;   
## [6] &quot;GSM1681155_5935446005_R05C01_Grn.idat.gz&quot;</code></pre>
<p>Currently <em><a href="http://bioconductor.org/packages/minfi">minfi</a></em> does not support reading compressed IDAT files. This is clearly a needed functionality and (as the maintainer of this package) I will address this. But for now we will need to decompress the files.</p>
<pre class="r"><code>idatFiles &lt;- list.files(&quot;GSE68777/idat&quot;, pattern = &quot;idat.gz$&quot;, full = TRUE)
head(sapply(idatFiles, gunzip, overwrite = TRUE), n = 3)</code></pre>
<pre><code>## GSE68777/idat/GSM1681154_5958091019_R03C02_Grn.idat.gz 
##                                                8091452 
## GSE68777/idat/GSM1681154_5958091019_R03C02_Red.idat.gz 
##                                                8091452 
## GSE68777/idat/GSM1681155_5935446005_R05C01_Grn.idat.gz 
##                                                8091452</code></pre>
<p>Now we read the IDAT files using <code>read.450k.exp()</code> which (in this case) reads all the IDAT files in a directory.</p>
<pre class="r"><code>rgSet &lt;- read.450k.exp(&quot;GSE68777/idat&quot;)
rgSet</code></pre>
<pre><code>## RGChannelSet (storageMode: lockedEnvironment)
## assayData: 622399 features, 40 samples 
##   element names: Green, Red 
## An object of class 'AnnotatedDataFrame': none
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19</code></pre>
<pre class="r"><code>pData(rgSet)</code></pre>
<pre><code>## data frame with 0 columns and 40 rows</code></pre>
<pre class="r"><code>head(sampleNames(rgSet))</code></pre>
<pre><code>## [1] &quot;GSM1681154_5958091019_R03C02&quot; &quot;GSM1681155_5935446005_R05C01&quot;
## [3] &quot;GSM1681156_5958091020_R01C01&quot; &quot;GSM1681157_5958091020_R03C02&quot;
## [5] &quot;GSM1681158_5935403032_R05C01&quot; &quot;GSM1681159_5958091019_R04C02&quot;</code></pre>
<p>Now we have the data, but note that we have no pheno data. And the filenames are very unhelpful here. These names consists of a GEO identifier (the <code>GSM</code> part) followed by a standard IDAT naming convention with a 10 digit number which is an array identifier followed by an identifier of the form <code>R01C01</code>. This is because each array actually allows for the hybridization of 12 samples in a 6x2 arrangement. The <code>5958091020_R01C0</code> means row 1 and column 1 on chip <code>5958091020</code>. This is all good, but does not help us understand which samples are cases and which are controls.</p>
<p>We now get the standard GEO representation to get the phenotype data stored in GEO. Most of the columns in this phenotype data are irrelevant (contains data such as the address of the person who submitted the data); we keep the useful ones. Then we clean it.</p>
<pre class="r"><code>geoMat &lt;- getGEO(&quot;GSE68777&quot;)
pD.all &lt;- pData(geoMat[[1]])
pD &lt;- pD.all[, c(&quot;title&quot;, &quot;geo_accession&quot;, &quot;characteristics_ch1.1&quot;, &quot;characteristics_ch1.2&quot;)]
head(pD)</code></pre>
<pre><code>##                        title geo_accession characteristics_ch1.1
## GSM1681154 5958091019_R03C02    GSM1681154      diagnosis: Mania
## GSM1681155 5935446005_R05C01    GSM1681155      diagnosis: Mania
## GSM1681156 5958091020_R01C01    GSM1681156        diagnosis: Ctr
## GSM1681157 5958091020_R03C02    GSM1681157        diagnosis: Ctr
## GSM1681158 5935403032_R05C01    GSM1681158      diagnosis: Mania
## GSM1681159 5958091019_R04C02    GSM1681159      diagnosis: Mania
##            characteristics_ch1.2
## GSM1681154           Sex: Female
## GSM1681155           Sex: Female
## GSM1681156             Sex: Male
## GSM1681157           Sex: Female
## GSM1681158           Sex: Female
## GSM1681159             Sex: Male</code></pre>
<pre class="r"><code>names(pD)[c(3,4)] &lt;- c(&quot;group&quot;, &quot;sex&quot;)
pD$group &lt;- sub(&quot;^diagnosis: &quot;, &quot;&quot;, pD$group)
pD$sex &lt;- sub(&quot;^Sex: &quot;, &quot;&quot;, pD$sex)</code></pre>
<p>We now need to merge this pheno data into the methylation data. To do so, we need a common sample identifier and we make sure we re-order the phenotype data in the same order as the methylation data. Finally we put the phenotype data inside the methylation data.</p>
<pre class="r"><code>sampleNames(rgSet) &lt;- sub(&quot;.*_5&quot;, &quot;5&quot;, sampleNames(rgSet))
rownames(pD) &lt;- pD$title
pD &lt;- pD[sampleNames(rgSet),]
pData(rgSet) &lt;- pD
rgSet</code></pre>
<pre><code>## RGChannelSet (storageMode: lockedEnvironment)
## assayData: 622399 features, 40 samples 
##   element names: Green, Red 
## An object of class 'AnnotatedDataFrame'
##   sampleNames: 5958091019_R03C02 5935446005_R05C01 ...
##     5935403032_R04C01 (40 total)
##   varLabels: title geo_accession group sex
##   varMetadata: labelDescription
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19</code></pre>
</div>
<div id="preprocessing" class="section level2">
<h2>Preprocessing</h2>
<p>The <code>rgSet</code> object is a class called <code>RGChannelSet</code> which represents two color data with a green and a red channel, very similar to an <code>ExpressionSet</code>.</p>
<p>The first step is usually to preprocess the data, using a number of functions including</p>
<ul>
<li><code>preprocessRaw()</code> : do nothing.</li>
<li><code>preprocessIllumina()</code> : use Illumina’s standard processing choices.</li>
<li><code>preprocessQuantile()</code> : use a version of quantile normalization adapted to methylation arrays.</li>
<li><code>preprocessNoob()</code> : use the NOOB background correction method.</li>
<li><code>preprocessSWAN()</code> : use the SWAN method.</li>
<li><code>preprocessFunnorm()</code> : use functional normalization.</li>
</ul>
<p>These functions output different types of objects.</p>
<p>The class hierarchy in minfi is as follows: data can be stored in an Methylation and Unmethylation channel or in a percent methylation (called Beta) channel. For the first case we have the class <code>MethylSet</code>, for the second case we have the class <code>RatioSet</code>. When you have methylation / unmethylation values you can still compute Beta values on the fly. You convert from a <code>MethylSet</code> to a <code>RatioSet</code> with <code>ratioConvert()</code>.</p>
<p>In addition to these two classes, we have <code>GenomicMethylSet</code> and <code>GenomicRatioSet</code>. The <code>Genomic</code> indicates that the data has been associated with genomic coordinates using the <code>mapToGenome()</code> function.</p>
<p>The starting point for most analyses ought to be a <code>GenomicRatioSet</code> class. If your preprocessing method of choice does not get you there, use <code>ratioConvert()</code> and <code>mapToGenome()</code> to go the last steps.</p>
<p>Let us run <code>preprocessQuantile()</code> which arrives at a <code>GenomicRatioSet</code>:</p>
<pre class="r"><code>grSet &lt;- preprocessQuantile(rgSet)</code></pre>
<pre><code>## [preprocessQuantile] Mapping to genome.
## [preprocessQuantile] Fixing outliers.
## [preprocessQuantile] Quantile normalizing.</code></pre>
<pre class="r"><code>grSet</code></pre>
<pre><code>## class: GenomicRatioSet 
## dim: 485512 40 
## metadata(0):
## assays(2): M CN
## rownames(485512): cg13869341 cg14008030 ... cg08265308 cg14273923
## rowData names(0):
## colnames(40): 5958091019_R03C02 5935446005_R05C01 ...
##   5958091020_R02C02 5935403032_R04C01
## colData names(5): title geo_accession group sex predictedSex
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19
## Preprocessing
##   Method: Raw (no normalization or bg correction)
##   minfi version: 1.18.2
##   Manifest version: 0.4.0</code></pre>
<p>This is like a <code>SummarizedExperiment</code>; we can get the location of the CpGs by</p>
<pre class="r"><code>granges(grSet)</code></pre>
<pre><code>## GRanges object with 485512 ranges and 0 metadata columns:
##              seqnames               ranges strand
##                 &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt;
##   cg13869341     chr1       [15865, 15865]      *
##   cg14008030     chr1       [18827, 18827]      *
##   cg12045430     chr1       [29407, 29407]      *
##   cg20826792     chr1       [29425, 29425]      *
##   cg00381604     chr1       [29435, 29435]      *
##          ...      ...                  ...    ...
##   cg17939569     chrY [27009430, 27009430]      *
##   cg13365400     chrY [27210334, 27210334]      *
##   cg21106100     chrY [28555536, 28555536]      *
##   cg08265308     chrY [28555550, 28555550]      *
##   cg14273923     chrY [28555912, 28555912]      *
##   -------
##   seqinfo: 24 sequences from hg19 genome; no seqlengths</code></pre>
<p>The usual methylation measure is called “Beta” values; equal to percent methylation and defined as <code>Meth</code> divided by <code>Meth + Unmeth</code>.</p>
<pre class="r"><code>getBeta(grSet)[1:3,1:3]</code></pre>
<pre><code>##            5958091019_R03C02 5935446005_R05C01 5958091020_R01C01
## cg13869341         0.7485333         0.7696497         0.7322275
## cg14008030         0.5300410         0.5893653         0.6044354
## cg12045430         0.0912596         0.1007678         0.1057603</code></pre>
<p>CpGs forms clusters known as “CpG Islands”. Areas close to CpG Islands are known as CpG Shores, followed by CpG Shelfs and finally CpG Open Sea probes. An easy way to get at this is to use</p>
<pre class="r"><code>head(getIslandStatus(grSet))</code></pre>
<pre><code>## [1] &quot;OpenSea&quot; &quot;OpenSea&quot; &quot;Island&quot;  &quot;Island&quot;  &quot;Island&quot;  &quot;OpenSea&quot;</code></pre>
</div>
<div id="differential-methylation" class="section level2">
<h2>Differential Methylation</h2>
<p>Once the data has been normalized, one possibility is to identify differentially methylated CpGs by using <em><a href="http://bioconductor.org/packages/limma">limma</a></em> on the Beta values.</p>
<p>Another possibility is to look for clusters of CpGs all changing in the same direction. One method for doing this is through the <code>bumphunter()</code> function which interfaces to the <em><a href="http://bioconductor.org/packages/bumphunter">bumphunter</a></em> package.</p>
</div>
<div id="other-resources" class="section level2">
<h2>Other Resources</h2>
<p>The vignette from the <a href="http://bioconductor.org/packages/minfi">minfi package</a>.</p>
</div>
<div id="sessioninfo" class="section level2">
<h2>SessionInfo</h2>
<p></p>
<pre><code>## R version 3.3.0 (2016-05-03)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: OS X 10.11.5 (El Capitan)
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] IlluminaHumanMethylation450kanno.ilmn12.hg19_0.2.1
##  [2] IlluminaHumanMethylation450kmanifest_0.4.0        
##  [3] GEOquery_2.38.4                                   
##  [4] minfi_1.18.2                                      
##  [5] bumphunter_1.12.0                                 
##  [6] locfit_1.5-9.1                                    
##  [7] iterators_1.0.8                                   
##  [8] foreach_1.4.3                                     
##  [9] Biostrings_2.40.0                                 
## [10] XVector_0.12.0                                    
## [11] SummarizedExperiment_1.2.2                        
## [12] GenomicRanges_1.24.0                              
## [13] GenomeInfoDb_1.8.1                                
## [14] IRanges_2.6.0                                     
## [15] S4Vectors_0.10.0                                  
## [16] lattice_0.20-33                                   
## [17] Biobase_2.32.0                                    
## [18] BiocGenerics_0.18.0                               
## [19] BiocStyle_2.0.2                                   
## [20] rmarkdown_0.9.6                                   
## 
## loaded via a namespace (and not attached):
##  [1] mclust_5.2              base64_2.0             
##  [3] Rcpp_0.12.5             Rsamtools_1.24.0       
##  [5] digest_0.6.9            chron_2.3-47           
##  [7] R6_2.1.2                plyr_1.8.3             
##  [9] RSQLite_1.0.0           evaluate_0.9           
## [11] httr_1.1.0              zlibbioc_1.18.0        
## [13] GenomicFeatures_1.24.2  data.table_1.9.6       
## [15] annotate_1.50.0         Matrix_1.2-6           
## [17] preprocessCore_1.34.0   splines_3.3.0          
## [19] BiocParallel_1.6.2      stringr_1.0.0          
## [21] RCurl_1.95-4.8          biomaRt_2.28.0         
## [23] rtracklayer_1.32.0      multtest_2.28.0        
## [25] pkgmaker_0.22           htmltools_0.3.5        
## [27] openssl_0.9.3           quadprog_1.5-5         
## [29] codetools_0.2-14        matrixStats_0.50.2     
## [31] XML_3.98-1.4            reshape_0.8.5          
## [33] GenomicAlignments_1.8.0 MASS_7.3-45            
## [35] bitops_1.0-6            grid_3.3.0             
## [37] nlme_3.1-127            xtable_1.8-2           
## [39] registry_0.3            DBI_0.4-1              
## [41] magrittr_1.5            formatR_1.4            
## [43] stringi_1.0-1           genefilter_1.54.2      
## [45] doRNG_1.6               limma_3.28.4           
## [47] nor1mix_1.2-1           RColorBrewer_1.1-2     
## [49] siggenes_1.46.0         tools_3.3.0            
## [51] illuminaio_0.14.0       rngtools_1.2.4         
## [53] survival_2.39-2         yaml_2.1.13            
## [55] AnnotationDbi_1.34.2    beanplot_1.2           
## [57] knitr_1.13</code></pre>
</div>
<div id="corrections" class="section level2">
<h2>Corrections</h2>
<p>Improvements and corrections to this document can be submitted on its <a href="https://github.com/kasperdanielhansen/genbioconductor/blob/master/Rmd/./back.Rmd">GitHub</a> in its <a href="https://github.com/kasperdanielhansen/genbioconductor">repository</a></p>
<div id="refs" class="references">

</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
